keps/sig-instrumentation/4785-resource-state-metrics/README.md

```diff
@@ -0,0 +1,1216 @@
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal targets the incubation and incorporation of a controller capable
+of essentially doing all that Kube State Metrics' Custom Resource State API
+offers, in addition to various benefits of its own owing to the controller
+lifecycle it is based upon.
+
+The controller aims to replace the existing Kube State Metrics' Custom Resource 
+State feature-set and be significantly much more maintainable and scalable than
+its predecessor while enabling the community the freedom to extend the supported
+set of DSLs to parse the configuration in a language that they are familiar
+with, instead of forcing them up a steep learning-curve for a self-defined DSL
+that they have to live with, as is the case with Kube State Metrics' Custom
+Resource State feature-set.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a cluster admin, I want to be able to express my Kube State Metrics' Custom 
+Resource State configurations in dedicated CRs, so I can benefit from everything
+the controller lifecycle has to offer, for e.g., not having to mount volumes and 
+redeploy their modifications, event emissions on metrics generation, 
+multiple managed resources to isolate configuration logic, etc.
+
+#### Story 2
+
+As a cluster admin, I want to be able to express my Kube State Metrics' Custom
+Resource State configurations in languages that are well-known throughout the
+ecosystem, and enables me to get going without learning another DSL 
+(domain-specific language).
+
+#### Story 3
+
+As a Cluster Admin, I want to expose additional metrics from core resources that
+Kube State Metrics does not provide out of the box, so I can observe custom
+tailored metrics for my use case.
+
+#### Story 4
+
+As a maintainer of a project that makes use of Custom Resources, I want to 
+provide configuration on how to generate metrics for them via an externally
+managed resource, so my users can deploy them along-side on the same cluster.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+The controller abides by the following principals:
+* Garbage in, garbage out: Invalid configurations will generate invalid metrics.
```

Comment:
Do you have an example of an invalid configuration that would produce invalid metrics? As far as I can tell since rsm is aware of the custom resource schemas, most if not all of the user inputs could be validated and instead of silencing configuration errors, rsm could surface them to the users

Reply:
Configuration errors are surfaced in the `self` metrics, logs, as well as the managed resource's `status` field.

My intention here was to mainly point out that, for instance, a label key that doesn't follow Prometheus' `[a-zA-Z_][a-zA-Z0-9_]*` format will not raise an error, and still be present in the generated metric, such as `{1="1"}`, which will be detected at ingestion. However, we still run some validations that make sense on the RSM-level, such as having the same number of label keys and values, which will be surfaced.